modules/imgproc/src/color_hsv.dispatch.cpp

```diff
@@ -274,38 +275,45 @@ bool oclCvtColorBGR2HSV( InputArray _src, OutputArray _dst, int bidx, bool full
 
     if(_src.depth() == CV_8U)
     {
-        static UMat sdiv_data;
-        static UMat hdiv_data180;
-        static UMat hdiv_data256;
-        static int sdiv_table[256];
-        static int hdiv_table180[256];
-        static int hdiv_table256[256];
-        static volatile bool initialized180 = false, initialized256 = false;
-        volatile bool & initialized = hrange == 180 ? initialized180 : initialized256;
+        static std::mutex mtx;
```

Comment:
Oh i understand. thx for the info. is the a list of features i should usually test for? i started to test for WITH_OPENCL=OFF and i will keep in mind that there are targets without thread support... are there more configurations i should keep in mind?

Reply:
Done 031abea. Tested locally by running opencv_test_improc with following build configuration. Btw. i found ```cv::utils::lock_guard``` but it doesn't seem necessary since there is C++11 support.

```shell
cmake -DINSTALL_BIN_EXAMPLES=OFF -DBUILD_SHARED_LIBS=ON -DWITH_OPENCL=ON -DBUILD_opencv_java=OFF -DBUILD_opencv_js=OFF -DBUILD_opencv_python2=OFF -DBUILD_opencv_python3=OFF -DBUILD_EXAMPLES=OFF -DBUILD_PACKAGE=OFF -DBUILD_TESTS=ON -DBUILD_PERF_TESTS=ON -DBUILD_DOCS=OFF -DBUILD_opencv_videoio=ON -DBUILD_opencv_highgui=ON -DBUILD_opencv_ts=ON -DBUILD_opencv_imgcodecs=ON -DBUILD_opencv_plot=OFF -DBUILD_opencv_tracking=OFF -DBUILD_opencv_video=OFF -DBUILD_opencv_world=ON -DCMAKE_C_FLAGS=-DOPENCV_DISABLE_THREAD_SUPPORT -DCMAKE_CXX_FLAGS=-DOPENCV_DISABLE_THREAD_SUPPORT ../opencv
```

Reply:
done: [8cbf9c7](https://github.com/opencv/opencv/pull/24444/commits/8cbf9c70e398cdf8357564ef0e5ab8df7f5d0eef)

Reply:
> `std::lock_guard` => `cv::AutoLock`
> 
> Did you measure performance changes due to added of `.copyTo()` calls?

I compared the current HEAD of 4.x with the branch of the PR (which i updated to the least 4.x HEAD) .
I used follwing build configuration for both branches:
```bash
cmake -DINSTALL_BIN_EXAMPLES=OFF -DBUILD_SHARED_LIBS=ON -DWITH_OPENCL=ON -DBUILD_opencv_java=OFF -DBUILD_opencv_js=OFF -DBUILD_opencv_python2=OFF -DBUILD_opencv_python3=OFF -DBUILD_EXAMPLES=OFF -DBUILD_PACKAGE=OFF -DBUILD_TESTS=ON -DBUILD_PERF_TESTS=ON -DBUILD_DOCS=OFF -DBUILD_opencv_videoio=ON -DBUILD_opencv_highgui=ON -DBUILD_opencv_ts=ON -DBUILD_opencv_imgcodecs=ON -DBUILD_opencv_plot=OFF -DBUILD_opencv_tracking=OFF -DBUILD_opencv_video=OFF -DBUILD_opencv_world=ON -DENABLE_PROFILING=ON ../opencv
```

And the following test program: https://github.com/kallaballa/OpenCV-Issues/blob/main/src/oclCvtColorBGR2HSV-race/perf-test.cpp

Without patch it takes: ~14.8s
With patch it takes: ~15.2s

So there is a slow-down. I'll optimize a bit and measure again.

Reply:
I made a local patch that avoids the copy. it is just slightly faster. {~15.1s}. I'll start profiling.

Reply:
I meassured also the variant no-static/no-thread_local/no-locking: 155.512s

Reply:
Please always reuse OpenCV performance tests or add them if something is missing.

Avoid using of hand-written tests cases

> suspicious results

Provided measurement code has bug due to async execution of `cvtColor(UMat)` (code should schedule OpenCL kernels, but do not wait the final result).

P.S. Mat to UMat coping is synchronous call in current API - forces wait of OpenCL execution queue.